### PR TITLE
Standard api

### PR DESCRIPTION
**Pull request by Mm7**
Initial attempt to solve #172. If you like this proposal I'll add support for more analysis. Suggetions and comments are welcome! Thank you.

### CODE REVIEW

**Comment by XVilka** [comment]
@Mm7 do you have a telegram? We have a dedicated radeco channel - would be awesome if you join it.

**Comment by XVilka** [comment]
You can join either telegram https://t.me/joinchat/AppM-AYGWGXzONimd3w0FA
or #radeco channel at freenode - both are fine and interconnected.

**Comment by XVilka** [comment]
You can join #radeco IRC channel on Freenode if you like.

**Comment by XVilka** [comment]
@chinmaydd have a look please too.

File: src/analysis/copy_propagation/mod.rs [hunk 1]
```diff
@@ -7,20 +11,61 @@ use petgraph::graph::NodeIndex;
...
+}
+
+impl FuncMutAnalyzer for CopyPropagation {
+    fn analyze(&mut self, func: &mut RadecoFunction) -> Box<AnalyzerResult> {
+        let mut result = CopyPropagationResult::new();
+        let ssa = func.ssa_mut();
+        loop {
+            let copies = CopyInfo::gather_copies(ssa);
+            if copies.is_empty() {
+                result.touched = false;
```
Comment (kriw) [inline src/analysis/copy_propagation/mod.rs]: I think this always returns `false` as AnalyzerResult

**Comment by chinmaydd** [comment]
Hi @Mm7, this is great work! I have some comments which I would like to hear your thoughts on:

* I feel that the `AnalyzerResult` trait should specify more requirements for an implementation. This could include mechanisms to verify/collect changes that were made during the analysis passes.
* Some analysis passes might inherently depend on the other for better/optimal results. I would like if you could incorporate a way to take such information into account, especially for the `run_all_analysis()` call. For example, it might be beneficial to run DCE after a high-level analysis pass (such as CSE) to take advantage of more optimizations. (I could be wrong here, in which case ignore it).

**Comment by Mm7** [comment]
> Hi @Mm7, this is great work! I have some comments which I would like to hear your thoughts on:
> 
>     * I feel that the `AnalyzerResult` trait should specify more requirements for an implementation. This could include mechanisms to verify/collect changes that were made during the analysis passes.
> 

To do that, I think that the best approach is splitting the job in two different passes:
1. The user calls `analyze` to get a list of `Change`s that can be applied to the IR (which is left untouched). `analyze` could also be made return some extra infos.
2. The `Change`s are inspected and applied as taste (by the user).

>     * Some analysis passes might inherently depend on the other for better/optimal results. I would like if you could incorporate a way to take such information into account, especially for the `run_all_analysis()` call. For example, it might be beneficial to run DCE after a high-level analysis pass (such as CSE) to take advantage of more optimizations. (I could be wrong here, in which case ignore it).

To do that you need analysis-specific knowledge (for example you need to know that DCE should be run after CSE). `run_all_analysis` was a "dump" way to run all the analysis without knowing which ones are available and when they should be called. Probably we should remove that function.

In my opinion, the user (aka the one that calls the analysis) should be implemented somewhere else (like in a new module). Here we can take care of calling the more appropriate one, apply useful `Changes`s and drop the useless ones.
`Analyzer` should only provide a uniform interface to all the possible analyzer passes.

**Comment by XVilka** [comment]
@Mm7 I agree on preserving also the original IR tree before the analysis pass. This way we can run two non-dependent analysis passes in parallel, maybe compare the results and apply only one or merge them into one.

**Comment by Mm7** [comment]
I'm trying to implement a graph-preserving version of `copy_propagation`. but I'm facing a problem: as soon as you start collecting changes to apply, you have no way to analyze the graph as it would look like after you applied all the previous changes, you can only see it in the original form!

For example, imagine that dce is called on the following IR. The `changes` vector is empty.
```
%1 = 10
%2 = 20
%3 = 30
return %2
```
The analyzer will notice that `%1` is unused, so a `removeNode(%1)` is pushed to the `changes` vec. Let's suppose that `removeNode` not only removes %1 but renames `%2 -> %1`, `%3 -> %2` (in order to keep ids small). The analyzer is oblivious of this because changes are applied by the user after the execution of the `analyze` func. If will then notice that `%3` is unsued, so again a `removeNode(%3)` is pushed.
When the user will try to apply the changes It will lead to wrong results (return of `%3` instead of `%2`, removal on non-existing `%3`).

You may argue that my example is useless because `removeNode` *does not* rename stuff, and you would probably be correct. But still, In case of complex changes (not just removing nodes), you could get some nasty side effects messing up the graph. Keeping track of all of this would be a nightmare, IMHO.

Here's few solutions I have in mind:

1. Clone the graph, work on the copy and then drop it when you are done. (I suppose this would be horrible performance-wise).
2. Have the ssa accessors know that we are working in a slightly different graph by passing the changes` as parameter. They will take care of returning the correct result. (I suppose this would be horrible to get working properly)
3. Make a list of what each change breaks of the graph and what is preserved. For example, If the change: `removeNode` breaks the property: `nodeIdsNotTouched`, then the analyzer will not be allowed to run IR accessors that would change their result in case of a change in node ids. (rust's borrow checker could be exploited to enforce these rules)

**Comment by XVilka** [comment]
While 1 and 2 look saner it will be a pain to run it on any sufficiently large programs indeed. So looks like 3 is the only viable choice. @kriw @chinmaydd @radare what do you think?

**Comment by Mm7** [comment]
@XVilka Maybe we could also add an indirection layer to access the graph: that layer should provide normal access while we are not inside an analysis, but during analysis it should collect all the changes applied to the graph; when an accessor is called it should dynamically patch the result according to the previously collected changes. This way `change`s could still be applied during the analysis pass without touching the graph.
This idea should be simpler to implement than 2. because it is at a "lower" level (aka graph storage level rather than ssa accessor level).

The access mechanism would look like:
graph wrapper

**Comment by kriw** [comment]
> While 1 and 2 look saner it will be a pain to run it on any sufficiently large programs indeed. So looks like 3 is the only viable choice. @kriw @chinmaydd @radare what do you think?

I think the combination of 1 and 3 is also viable.
We can reduce the number of clones with 3 and possibly a few clones are acceptable.

**Comment by chinmaydd** [comment]
Sorry for my delay in response.

> run_all_analysis was a "dump" way to run all the analysis without knowing which ones are available and when they should be called. Probably we should remove that function.

Agree. I do not see any consumer interested in calling `run_all_analysis()` since it does not really provide any control over order and choice.

> Maybe we could also add an indirection layer to access the graph

The design looks good, but I agree with @kriw. How much performance hit due to cloning are we talking here ? Dont think it should be too much.

**Comment by Mm7** [comment]
@XVilka @chinmaydd @kriw I just uploaded a new version.
1. I added an `Action` enum to store "atomic" changes to the IR. For now only "replaceValue" is implemented. ASAP I'll add one variant for each one of the `SSAMod` accessors.
2. For now I just used the `clone` way.
3. How should we handle dependencies among analyzers? (for example CSE needs SSASort to be run) Should we run the needed analyzers on entry? Or require the user to honor dependencies before calling the analyzer (In this case how can we enforce/ensure this is done?)? Maybe a third way is simply making analyzers independent of each other.
4. There is some [redundancy`](https://github.com/radareorg/radeco-lib/blob/3b5f96cbe3d3ec8a329b3ae8847046a84d30a89f/src/analysis/cse/cse.rs#L118:L121) when applying the change to the cloned graph and pushing the change to the vector. I think that a wrapper would be useful here; furthermore it could also provide some logic to avoid clones when not needed.

**Comment by XVilka** [comment]
3. For dependencies - I think not to run, but to check if analysis stage was running before. Of course making them independent is better. The goal is to be able to run them multiple times, like LLVM passes for example.

**Comment by XVilka** [comment]
@Mm7 ping?

**Comment by Mm7** [comment]
Sorry for the delay. Currently I'm studying hard for my exams. But I'm still working on this

**Comment by XVilka** [comment]
@Mm7 ah, ok then. Good luck with your exams!

**Comment by XVilka** [comment]
@Mm7 well, if it is too hard to implement, then it is probably a wrong solution. Lets proceed then with mutable transformations then.

**Comment by XVilka** [comment]
@kriw @chinmaydd please take a look again to the proposed API.

**Comment by kriw** [comment]
I think this is a good starting point for a standard api.
LGTM :)

**Comment by Mm7** [comment]
OK, thanks for you feedback. Next week my exams will be over so I'll have more time to finish.

**Comment by XVilka** [comment]
Looks good, with this API code looks much clearer, good job!

**Comment by kriw** [comment]
LGTM :)

**Pull request merged**